File: examples/01-foundations/python/deterministic_replay_good.py
from datetime import timedelta

from temporalio import workflow

# `load` is a module-level @activity.defn function (does the actual I/O).


@workflow.defn
class ReplaySafeWorkflow:
    @workflow.run
    async def run(self, batch_date: str) -> None:
        # Temporal records these deterministic decisions in Workflow history.
        now = workflow.now()                 # replay-safe time source
        shard = workflow.random().randint(0, 9)  # replay-safe, seeded RNG

        # I/O belongs in Activities because Activity results are recorded.
        await workflow.execute_activity(
            load,
            f"s3://bucket/clean/{batch_date}/shard={shard}?ts={now.timestamp()}",
            start_to_close_timeout=timedelta(minutes=10),
        )

File: examples/01-foundations/go/deterministic_replay_good.go
package foundations

import (
	"fmt"
	"time"

	"go.temporal.io/sdk/workflow"
)

func ReplaySafeWorkflow(ctx workflow.Context, batchDate string) error {
	ctx = workflow.WithActivityOptions(ctx, workflow.ActivityOptions{
		StartToCloseTimeout: 10 * time.Minute,
	})

	// Temporal records these deterministic decisions in Workflow history.
	now := workflow.Now(ctx) // replay-safe time source

	// Replay-safe randomness: use a side effect so the value is recorded once.
	var shard int
	_ = workflow.SideEffect(ctx, func(workflow.Context) any {
		return rand10()
	}).Get(&shard)

	// I/O belongs in Activities because Activity results are recorded in history.
	uri := fmt.Sprintf("s3://bucket/clean/%s/shard=%d?ts=%d", batchDate, shard, now.Unix())
	return workflow.ExecuteActivity(ctx, "Load", uri).Get(ctx, nil)
}
